<h1 style="display: inline;" >6. Grid Search and Stacking (B2)


The performance of most machine learning models depends on the settings of several hyperparameters. Optimizing model performance is generally a manner of tuning these hyperparameters and testing performance. 

Testing performance on a subspace of parameter settings is known as a 'grid search'. In its default state, the user supplies a list of possible hyperparameter settings, and the search is conducted over all possibilities. 

Since this can be quite time-consuming, H2O allows you to stipulate the 'max_models' (i.e. maximum number of parameter combinations to consider), with each model chosen randomly from among the given possibilities (i.e. a "RandomDiscrete" search strategy).

Here we import the airline data and create a number of different models by random search of hyperparameter space.

**Importing the data:**

In [1]:
import h2o
h2o.init()

[WARNING] H2O requires requests module of version 2.10 or newer. You have version 2.6.0.
You can upgrade to the newest version of the module running from the command line
    $ pip2 install --upgrade requests
Checking whether there is an H2O instance running at http://localhost:54321. connected.


--------------------------  ----------------------------------------
H2O cluster uptime:         59 mins 52 secs
H2O cluster version:        3.14.0.6
H2O cluster version age:    10 days
H2O cluster name:           ec2-user
H2O cluster total nodes:    2
H2O cluster free memory:    966 Mb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         locked, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             2.7.5 final
--------------------------  ----------------------------------------

In [3]:
flights = h2o.import_file("hdfs://ec2-34-204-73-232.compute-1.amazonaws.com:9000/allyears2k.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


Our goal is to predict whether the arrival is delayed, "ArrDelay". We will use date and time, scheduled time, origin and destination, and airline.


In [4]:
pred = flights[['IsArrDelayed', 'Year','Month','DayofMonth','DayOfWeek','CRSDepTime','CRSArrTime','CRSElapsedTime','UniqueCarrier','FlightNum','TailNum','Origin','Dest','Distance']]
pred_train, pred_valid = pred.split_frame(ratios=[0.8])

# Building Models
** Generalized Linear Models:**

In [5]:
from h2o.grid.grid_search import H2OGridSearch
from h2o.estimators.glm import H2OGeneralizedLinearEstimator
hyper_parameters = {'alpha': [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0], 'lambda': [0, 1e-7, 1e-5, 1e-3, 1e-1]}
criteria = {"strategy": "RandomDiscrete", "max_models": 5, "seed": 1234}
gs1 = H2OGridSearch(H2OGeneralizedLinearEstimator(family='binomial', nfolds=2, fold_assignment="modulo", keep_cross_validation_predictions=True), hyper_parameters, search_criteria=criteria)

x_cols = ['Year','Month','DayofMonth','DayOfWeek','CRSDepTime','CRSArrTime','CRSElapsedTime','UniqueCarrier','FlightNum','TailNum','Origin','Dest','Distance']

gs1.train(x=x_cols, y="IsArrDelayed", training_frame=pred_train, validation_frame=pred_valid)
auc_glm = gs1.auc(valid=True)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


**Gradient Boosted Machines:**

In [6]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.grid.grid_search import H2OGridSearch
hyper_parameters = {'learn_rate': [0.01, 0.03], 'max_depth': [3,4,5,6,9], 'sample_rate': [0.7,0.8,0.9,1.0], 'col_sample_rate': [0.2,0.3,0.4,0.5,0.6,0.7,0.8]}
criteria = {"strategy": "RandomDiscrete", "max_models": 5, "seed": 1234}
gs2 = H2OGridSearch(H2OGradientBoostingEstimator(nfolds=2, fold_assignment="modulo", keep_cross_validation_predictions=True), hyper_parameters, search_criteria=criteria)

x_cols = ['Year','Month','DayofMonth','DayOfWeek','CRSDepTime','CRSArrTime','CRSElapsedTime','UniqueCarrier','FlightNum','TailNum','Origin','Dest','Distance']

gs2.train(x=x_cols, y="IsArrDelayed", training_frame=pred_train, validation_frame=pred_valid)
auc_gbm = gs2.auc(valid=True)

gbm Grid Build progress: |████████████████████████████████████████████████| 100%


**Multi-layer perceptrons:**

In [7]:
from h2o.estimators.deeplearning import H2ODeepLearningEstimator
from h2o.grid.grid_search import H2OGridSearch
hyper_parameters = {'activation': ["rectifier", "rectifier_with_dropout"], 'hidden': [[10, 10],[20, 15],[50,50,50]], 'l1': [0, 1e-3, 1e-5], 'l2': [0, 1e-3, 1e-5]}
criteria = {"strategy": "RandomDiscrete", "max_models": 5, "seed": 1234}
gs3 = H2OGridSearch(H2ODeepLearningEstimator(nfolds=2, fold_assignment="modulo", keep_cross_validation_predictions=True), hyper_parameters, search_criteria=criteria)
gs3.train(x=x_cols, y="IsArrDelayed", training_frame=pred_train, validation_frame=pred_valid)
auc_mlp = gs3.auc()

deeplearning Grid Build progress: |███████████████████████████████████████| 100%


# Evaluating Model Performance

In [8]:
auc = auc_glm
auc.update(auc_gbm)
auc.update(auc_mlp)
auc

{u'Grid_DeepLearning_py_4_sid_93ae_model_python_1508440543160_871_model_0': 0.7100269514556684,
 u'Grid_DeepLearning_py_4_sid_93ae_model_python_1508440543160_871_model_1': 0.7071959224861313,
 u'Grid_DeepLearning_py_4_sid_93ae_model_python_1508440543160_871_model_2': 0.6688345482458213,
 u'Grid_DeepLearning_py_4_sid_93ae_model_python_1508440543160_871_model_3': 0.7042174007318446,
 u'Grid_DeepLearning_py_4_sid_93ae_model_python_1508440543160_871_model_4': 0.6565614708058285,
 u'Grid_GBM_py_4_sid_93ae_model_python_1508440543160_47_model_0': 0.6830318707483571,
 u'Grid_GBM_py_4_sid_93ae_model_python_1508440543160_47_model_1': 0.6797463558257494,
 u'Grid_GBM_py_4_sid_93ae_model_python_1508440543160_47_model_2': 0.6922829123052446,
 u'Grid_GBM_py_4_sid_93ae_model_python_1508440543160_47_model_3': 0.6793678034410898,
 u'Grid_GBM_py_4_sid_93ae_model_python_1508440543160_47_model_4': 0.7132884927724197,
 u'Grid_GLM_py_4_sid_93ae_model_python_1508440543160_1_model_0': 0.6532549396398448,
 u'Gr

# Stacking

Next, we create a composite model as an ensemble of smaller models — a process known as stacking. The H2O stacking model creates an ensemble by fitting a GLM on top of the results of the individual models.

In [9]:
models = list(auc.keys())

In [10]:
from h2o.estimators.stackedensemble import H2OStackedEnsembleEstimator

stack = H2OStackedEnsembleEstimator(base_models=models)
stack.train(x=x_cols, y="IsArrDelayed", training_frame=pred_train, validation_frame=pred_valid)

stackedensemble Model Build progress: |███████████████████████████████████| 100%

ModelMetricsBinomialGLM: stackedensemble
** Reported on train data. **

MSE: 0.198670047585
RMSE: 0.445724183308
LogLoss: 0.581557336202
Null degrees of freedom: 35243
Residual degrees of freedom: 35238
Null deviance: 48422.9937834
Residual deviance: 40992.8135142
AIC: 41004.8135142
AUC: 0.762249457365
Gini: 0.524498914731
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.414364127528: 


NO    YES    Error    Rate
-----  ----  -----  -------  -----------------
NO     6562  9103   0.5811   (9103.0/15665.0)
YES    2269  17310  0.1159   (2269.0/19579.0)
Total  8831  26413  0.3227   (11372.0/35244.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.414364     0.75274   273
max f2                       0.256609     0.866412  357
max f0point5                 0.558816     0.729476  186
max accuracy                 0.517943     0.694728  212
max precision                0.915047     1         0
max recall                   0.125984     1         397
max specificity              0.915047     1         0
max absolute_mcc             0.556937     0.381749  187
max min_per_class_accuracy   0.542948     0.690229  196
max mean_per_class_accuracy  0.556937     0.692046  187

Gains/Lift Table: Avg response rate: 55.55 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0100159                   0.901734           1.78479   1.78479            0.991501         0.991501                    0.0178763       0.0178763                  78.4794   78.4794
    2        0.0200034                   0.89074            1.71316   1.74903            0.951705         0.971631                    0.0171102       0.0349865                  71.3156   74.9025
    3        0.0300193                   0.877244           1.7185    1.73884            0.954674         0.965974                    0.0172123       0.0521988                  71.8501   73.8841
    4        0.0400068                   0.860421           1.69781   1.7286             0.943182         0.960284                    0.0169569       0.0691557                  69.7814   72.8599
    5        0.0500227                   0.84406            1.64201   1.71126            0.912181         0.950652                    0.0164462       0.0856019                  64.201    71.1262
    6        0.100017                    0.779721           1.61109   1.66119            0.895006         0.922837                    0.0805455       0.166147                   61.1092   66.1191
    7        0.150011                    0.737769           1.52221   1.61487            0.84563          0.897106                    0.0761019       0.242249                   52.2212   61.4873
    8        0.200006                    0.707982           1.4272    1.56796            0.792849         0.871046                    0.071352        0.313601                   42.7201   56.7962
    9        0.299994                    0.655529           1.2949    1.47695            0.719353         0.820486                    0.129475        0.443077                   29.4902   47.695
    10       0.400011                    0.603776           1.18985   1.40516            0.660993         0.780607                    0.119005        0.562082                   18.9848   40.5165
    11       0.5                         0.552397           1.08394   1.34093            0.602157         0.744921                    0.108381        0.670463                   8.39373   34.0927
    12       0.599989                    0.500711           0.938357  1.27384            0.521283         0.707652                    0.093825        0.764288                   -6.16433  27.3838
    13       0.700006                    0.444534           0.830851  1.21054            0.46156          0.67249                     0.0830992       0.847388                   -16.9149  21.0544
    14       0.799994                    0.382049           0.687038  1.14511            0.381669         0.636141                    0.0686961       0.916084                   -31.2962  14.5113
    15       0.899983                    0.309055           0.516939  1.07532            0.287174         0.597371                    0.051688        0.967772                   -48.3061  7.53221
    16       1                           0.103682           0.322229  1                  0.179007         0.555527                    0.0322284       1                          -67.7771  0

Here we can compare the performance of the stacked model to that of the individual models above. Performance varies due to the randomness of our grid search, but on last run the AUC improved from 0.713 (best individual model) to 0.762 (stacked model):

In [ ]:
stack.model_performance()